File: packages/sdk-py/src/agent_relay/communicate/adapters/google_adk.py
"""Google ADK adapter for on_relay()."""

from __future__ import annotations
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..core import Relay

def on_relay(agent: Any, relay: Relay) -> Any:
    """Wrap Google ADK Agent to connect it to the relay."""
    
    # 1. Add tools for sending
    async def relay_send(to: str, text: str) -> str:
        """Send a private message to another agent."""
        await relay.send(to, text)
        return "Message sent"

    async def relay_inbox() -> str:
        """Check for new messages in the inbox."""
        messages = await relay.inbox()
        if not messages: return "No new messages"
        return "\n".join([f"From {m.sender}: {m.text}" for m in messages])

    async def relay_post(channel: str, text: str) -> str:
        """Post a message to a shared channel."""
        await relay.post(channel, text)
        return "Message posted"

    async def relay_agents() -> str:
        """List all agents currently on the relay."""
        agents = await relay.agents()
        return ", ".join(agents)

    agent.tools.extend([relay_send, relay_inbox, relay_post, relay_agents])

    # 2. Inject before_model_callback for receiving
    orig_callback = getattr(agent, "before_model_callback", None)

    async def relay_callback(llm_request: Any) -> None:
        if orig_callback:
            await orig_callback(llm_request)
        
        messages = await relay.inbox()
        if messages:
            content = "\n\nNew messages from other agents:\n"
            for m in messages:
                content += f"  Relay message from {m.sender}: {m.text}\n"
            
            # Assuming llm_request.contents is a list of user messages
            # and we can append a new user part/content
            llm_request.contents.append(content)

    agent.before_model_callback = relay_callback
    return agent

File: packages/sdk/src/communicate/adapters/google-adk.ts
import { Relay } from '../core.js';
import { formatRelayInbox, formatRelayMessage, type Message, type MessageCallback } from '../types.js';

type JsonObjectSchema = {
  type: 'object';
  properties: Record<string, Record<string, unknown>>;
  required: string[];
};

type FunctionToolLike = {
  name: string;
  description: string;
};

type ToolUnionLike = FunctionToolLike;

type RunnerLike = {
  runAsync(params: {
    userId: string;
    sessionId: string;
    newMessage: { role: string; parts: Array<{ text: string }> };
  }): AsyncGenerator<unknown, void, undefined>;
};

type LlmAgentLike = {
  name: string;
  tools: ToolUnionLike[];
};

type RelayLike = {
  send(to: string, text: string): Promise<void>;
  post(channel: string, text: string): Promise<void>;
  inbox(): Promise<Message[]>;
  agents(): Promise<string[]>;
  onMessage(callback: MessageCallback): () => void;
};

interface GoogleAdkRelayOptions {
  /** Google ADK LlmAgent instance. */
  agent: LlmAgentLike;
  /** Optional Runner for routing incoming messages to the agent. */
  runner?: RunnerLike;
  /** User ID for runner sessions. Defaults to 'relay'. */
  userId?: string;
  /** Session ID for runner sessions. Defaults to 'relay-session'. */
  sessionId?: string;
}

/**
 * Create a FunctionTool-shaped object for Google ADK.
 * We construct the tool objects duck-typed to avoid importing the actual
 * FunctionTool class (which requires @google/genai at runtime).
 */
function createRelayFunctionTools(relay: RelayLike): FunctionToolLike[] {
  // We dynamically import FunctionTool to create real ADK tool instances
  // For now, we build tool-like objects that match the FunctionTool constructor shape
  // and lazily construct them.
  const toolDefs: Array<{
    name: string;
    description: string;
    parameters: JsonObjectSchema;
    execute: (input: Record<string, string>) => Promise<Record<string, unknown>>;
  }> = [
    {
      name: 'relay_send',
      description: 'Send a direct message to another relay agent.',
      parameters: { type: 'object', properties: { to: { type: 'string' }, text: { type: 'string' } }, required: ['to', 'text'] },
      async execute(input) {
        await relay.send(input.to, input.text);
        return { result: `Sent relay message to ${input.to}.` };
      },
    },
    {
      name: 'relay_inbox',
      description: 'Drain and inspect newly received relay messages.',
      parameters: { type: 'object', properties: {}, required: [] },
      async execute() {
        return { result: formatRelayInbox(await relay.inbox()) };
      },
    },
    {
      name: 'relay_post',
      description: 'Post a message to a relay channel.',
      parameters: { type: 'object', properties: { channel: { type: 'string' }, text: { type: 'string' } }, required: ['channel', 'text'] },
      async execute(input) {
        await relay.post(input.channel, input.text);
        return { result: `Posted relay message to #${input.channel}.` };
      },
    },
    {
      name: 'relay_agents',
      description: 'List currently online relay agents.',
      parameters: { type: 'object', properties: {}, required: [] },
      async execute() {
        return { result: (await relay.agents()).join('\n') };
      },
    },
  ];

  return toolDefs;
}

/**
 * Attach relay communication tools and message routing to a Google ADK agent.
 * @param name - Agent name for relay registration.
 * @param options - Google ADK agent and optional runner configuration.
 * @param relay - Optional pre-configured Relay instance.
 * @returns Object with the augmented agent, an unsubscribe function, and the relay tools added.
 */
export function onRelay(
  name: string,
  options: GoogleAdkRelayOptions,
  relay: RelayLike = new Relay(name),
): { agent: LlmAgentLike; tools: FunctionToolLike[]; unsubscribe: () => void } {
  const { agent, runner, userId = 'relay', sessionId = 'relay-session' } = options;

  const relayTools = createRelayFunctionTools(relay);

  // Append relay tools to the agent's tool list
  for (const tool of relayTools) {
    agent.tools.push(tool as ToolUnionLike);
  }

  let unsubscribe: () => void = () => {};

  if (runner) {
    unsubscribe = relay.onMessage(async (message: Message) => {
      const prompt = formatRelayMessage(message);
      const content = { role: 'user', parts: [{ text: prompt }] };

      // Drain the async generator to completion
      for await (const _event of runner.runAsync({ userId, sessionId, newMessage: content })) {
        // Events are consumed; the agent handles them internally.
      }
    });
  }

  return { agent, tools: relayTools, unsubscribe };
}
